File: src_all/sources/MobileIMSDK4jXOpen/src/net/openmob/mobileimsdk/java/ClientCoreSDK.java
/*
 * Copyright (C) 2017  即时通讯网(52im.net) & Jack Jiang.
 * The MobileIMSDK_X (MobileIMSDK v3.x) Project. 
 * All rights reserved.
 * 
 * > Github地址: https://github.com/JackJiang2011/MobileIMSDK
 * > 文档地址: http://www.52im.net/forum-89-1.html
 * > 即时通讯技术社区：http://www.52im.net/
 * > 即时通讯技术交流群：320837163 (http://www.52im.net/topic-qqgroup.html)
 *  
 * "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
 * 
 * ClientCoreSDK.java at 2017-5-1 22:14:56, code by Jack Jiang.
 * You can contact author with jack.jiang@52im.net or jb2011@163.com.
 */
package net.openmob.mobileimsdk.java;

import net.openmob.mobileimsdk.java.core.AutoReLoginDaemon;
import net.openmob.mobileimsdk.java.core.KeepAliveDaemon;
import net.openmob.mobileimsdk.java.core.LocalUDPDataReciever;
import net.openmob.mobileimsdk.java.core.LocalUDPSocketProvider;
import net.openmob.mobileimsdk.java.core.QoS4ReciveDaemon;
import net.openmob.mobileimsdk.java.core.QoS4SendDaemon;
import net.openmob.mobileimsdk.java.event.ChatBaseEvent;
import net.openmob.mobileimsdk.java.event.ChatTransDataEvent;
import net.openmob.mobileimsdk.java.event.MessageQoSEvent;

public class ClientCoreSDK
{
	private final static String TAG = ClientCoreSDK.class.getSimpleName();
	
	public static boolean DEBUG = true;
	public static boolean autoReLogin = true;
	
	private static ClientCoreSDK instance = null;
	
	private boolean _init = false;
	private boolean connectedToServer = true;
	private boolean loginHasInit = false;
	private String currentLoginUserId = null;
	private String currentLoginToken = null;
	private String currentLoginExtra = null;
	
	private ChatBaseEvent chatBaseEvent = null;
	private ChatTransDataEvent chatTransDataEvent = null;
	private MessageQoSEvent messageQoSEvent = null;
	
	public static ClientCoreSDK getInstance()
	{
		if(instance == null)
			instance = new ClientCoreSDK();
		return instance;
	}
	
	private ClientCoreSDK()
	{
	}
	
	public void init()
	{
		if(!_init)
		{
			_init = true;
		}
	}
	
	public void release()
	{
		LocalUDPSocketProvider.getInstance().closeLocalUDPSocket();
	    AutoReLoginDaemon.getInstance().stop(); // 2014-11-08 add by Jack Jiang
		QoS4SendDaemon.getInstance().stop();
		KeepAliveDaemon.getInstance().stop();
		LocalUDPDataReciever.getInstance().stop();
		QoS4ReciveDaemon.getInstance().stop();
		
		_init = false;
		
		this.setLoginHasInit(false);
		this.setConnectedToServer(false);
	}
	
	public String getCurrentLoginUserId()
	{
		return currentLoginUserId;
	}
	public ClientCoreSDK setCurrentLoginUserId(String currentLoginUserId)
	{
		this.currentLoginUserId = currentLoginUserId;
		return this;
	}
	
	public String getCurrentLoginToken()
	{
		return currentLoginToken;
	}
	public void setCurrentLoginToken(String currentLoginToken)
	{
		this.currentLoginToken = currentLoginToken;
	}
	
	public String getCurrentLoginExtra()
	{
		return currentLoginExtra;
	}
	public ClientCoreSDK setCurrentLoginExtra(String currentLoginExtra)
	{
		this.currentLoginExtra = currentLoginExtra;
		return this;
	}

	public boolean isLoginHasInit()
	{
		return loginHasInit;
	}
	public ClientCoreSDK setLoginHasInit(boolean loginHasInit)
	{
		this.loginHasInit = loginHasInit;
		return this;
	}
	
	public boolean isConnectedToServer()
	{
		return connectedToServer;
	}
	public void setConnectedToServer(boolean connectedToServer)
	{
		this.connectedToServer = connectedToServer;
	}

	public boolean isInitialed()
	{
		return this._init;
	}

	public void setChatBaseEvent(ChatBaseEvent chatBaseEvent)
	{
		this.chatBaseEvent = chatBaseEvent;
	}
	public ChatBaseEvent getChatBaseEvent()
	{
		return chatBaseEvent;
	}
	
	public void setChatTransDataEvent(ChatTransDataEvent chatTransDataEvent)
	{
		this.chatTransDataEvent = chatTransDataEvent;
	}
	public ChatTransDataEvent getChatTransDataEvent()
	{
		return chatTransDataEvent;
	}
	
	public void setMessageQoSEvent(MessageQoSEvent messageQoSEvent)
	{
		this.messageQoSEvent = messageQoSEvent;
	}
	public MessageQoSEvent getMessageQoSEvent()
	{
		return messageQoSEvent;
	}
}

File: src_all/libs_src/MobileIMSDK4i_x_open/MobileIMSDK4i/Core/ClientCoreSDK.m
//  ----------------------------------------------------------------------
//  Copyright (C) 2017  即时通讯网(52im.net) & Jack Jiang.
//  The MobileIMSDK_X (MobileIMSDK v3.x) Project.
//  All rights reserved.
//
//  > Github地址: https://github.com/JackJiang2011/MobileIMSDK
//  > 文档地址: http://www.52im.net/forum-89-1.html
//  > 即时通讯技术社区：http://www.52im.net/
//  > 即时通讯技术交流群：320837163 (http://www.52im.net/topic-qqgroup.html)
//
//  "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
//
//  如需联系作者，请发邮件至 jack.jiang@52im.net 或 jb2011@163.com.
//  ----------------------------------------------------------------------
//
//  ClientCoreSDK.m
//  MibileIMSDK4i_X (MobileIMSDK v3.0 at Summer 2017)
//
//  Created by JackJiang on 14/10/21.
//  Copyright (c) 2017年 52im.net. All rights reserved.
//

#import "ClientCoreSDK.h"
#import "ChatTransDataEvent.h"
#import "ChatBaseEvent.h"
#import "MessageQoSEvent.h"
#import "Reachability.h"
#import "QoS4SendDaemon.h"
#import "KeepAliveDaemon.h"
#import "LocalUDPDataReciever.h"
#import "LocalUDPSocketProvider.h"
#import "QoS4ReciveDaemon.h"
#import "AutoReLoginDaemon.h"


static BOOL ENABLED_DEBUG = NO;
static BOOL autoReLogin = YES;


@interface ClientCoreSDK ()

@property (nonatomic) BOOL _init;
@property (nonatomic) Reachability *internetReachability;

@end


@implementation ClientCoreSDK

static ClientCoreSDK *instance = nil;

+ (ClientCoreSDK *)sharedInstance
{
    if (instance == nil)
    {
        instance = [[super allocWithZone:NULL] init];
    }
    return instance;
}

+ (BOOL) isENABLED_DEBUG
{
    return ENABLED_DEBUG;
}
+ (void) setENABLED_DEBUG:(BOOL)enabledDebug
{
    ENABLED_DEBUG = enabledDebug;
}

+ (BOOL) isAutoReLogin
{
    return autoReLogin;
}
+ (void) setAutoReLogin:(BOOL)arl
{
    autoReLogin = arl;
}

- (id)init
{
    if (![super init])
        return nil;
    
    return self;
}

- (void)initCore
{
    if(!self._init)
    {
        self.localDeviceNetworkOk = NO;
        self.connectedToServer = NO;
        self.loginHasInit = NO;
        
        if(self.internetReachability == nil)
        {
            [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(reachabilityChanged:) name:kReachabilityChangedNotification object:nil];
            self.internetReachability = [Reachability reachabilityForInternetConnection];
        }
        [self.internetReachability startNotifier];
        self.localDeviceNetworkOk = [self internetReachable];
        self._init = YES;
        
        NSLog(@"ClientCoreSDK已经完成initCore了！");
    }
}

- (void) releaseCore
{
    [[AutoReLoginDaemon sharedInstance] stop]; 
    [[QoS4ReciveDaemon sharedInstance] stop];
    [[KeepAliveDaemon sharedInstance] stop];
    [[QoS4SendDaemon sharedInstance] stop];
    [[LocalUDPSocketProvider sharedInstance] closeLocalUDPSocket];

    [[QoS4SendDaemon sharedInstance] clear];
    [[QoS4ReciveDaemon sharedInstance] clear];

    [self.internetReachability stopNotifier];
    
    self._init = NO;
    self.loginHasInit = NO;
    self.connectedToServer = NO;
}


- (BOOL) isInitialed
{
    return self._init;
}

- (BOOL)internetReachable
{
    NetworkStatus netStatus = [self.internetReachability currentReachabilityStatus];
    return netStatus == ReachableViaWWAN || netStatus == ReachableViaWiFi;
}

- (void) reachabilityChanged:(NSNotification *)note
{
    Reachability* reachability = [note object];
    NSParameterAssert([reachability isKindOfClass:[Reachability class]]);
    
    NetworkStatus netStatus = [reachability currentReachabilityStatus];
    BOOL connectionRequired = [reachability connectionRequired];
    NSString* statusString = @"";
    
    switch (netStatus)
    {
        case NotReachable:
        {
            statusString = NSLocalizedString(@"【IMCORE】【本地网络通知】检测本地网络连接断开了!", @"Text field text for access is not available");
            connectionRequired = NO;
            
            self.localDeviceNetworkOk = false;
            [[LocalUDPSocketProvider sharedInstance] closeLocalUDPSocket];
            
            break;
        }
            
        case ReachableViaWWAN: // 蜂窝网络、3G网络等
        case ReachableViaWiFi: // WIFI
        {
            int wifi = (netStatus == ReachableViaWiFi);
            statusString= [NSString stringWithFormat:NSLocalizedString(@"【IMCORE】【本地网络通知】检测本地网络已连接上了! WIFI? %d", @""), wifi?@"YES":@"NO"];
            
            self.localDeviceNetworkOk = true;
            [[LocalUDPSocketProvider sharedInstance] closeLocalUDPSocket];
            
            break;
        }
    }
    
    if (connectionRequired)
    {
        NSString *connectionRequiredFormatString = NSLocalizedString(@"【IMCORE】%@, Connection Required", @"Concatenation of status string with connection requirement");
        statusString = [NSString stringWithFormat:connectionRequiredFormatString, statusString];
    }
    
    if(ENABLED_DEBUG)
        NSLog(@"%@", statusString);
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:kReachabilityChangedNotification object:nil];
}

@end
